torch/nn/utils/rnn.py

```diff
@@ -123,3 +124,126 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
     return output, lengths
+
+
+def pad_sequence(sequences, lengths, batch_first=False):
+    r"""Pad a list of variable length Variables with zero
+
+    The ``pad_sequence`` pads the list of Variables on zeroth dimension and
+    stack all the sequences on zeroth dimension. For example, if the input is
+    list of sequences with size `` Lx*`` and if batch_first is False, the
+    output will be of size `` TxBx* `` and if batch_first is True,
+    output will be of size ``BxTx* ``. The ``pad_sequence`` accepts list of
+    sequences and its lengths which should be sorted in decreasing order
+
+    B is batch size - Number of elements in ``sequences``
+    T is length longest sequence
+    L is length of the sequence
+    * is any number of trailing dimensions, including none
+
+    >>> from torch.nn.utils.rnn import pad_sequence
+    >>> a = Variable(torch.ones(25, 300))
+    >>> b = Variable(torch.ones(22, 300))
+    >>> c = Variable(torch.ones(15, 300))
+    >>> pad_sequence([a, b, c], [25, 22, 15]).size()
+    torch.Size([25, 3, 300])
+
+    Note:
+        This function returns a Variable of size TxBx* or BxTx* where T is the
+            length of longest sequence (lengths[0])
+        Function assumes trailing dimensions and type of all the Variables
+            in sequences are same
+
+    Arguments:
+        sequences (list[Variable]): list of variable length sequences.
+        lengths (list[int]): list of sequences lengths of each batch element.
+        batch_first (bool, optional): if True, the input is expected in Bx*x*
+            format.
+
+    Returns:
+        Variable of size ``T x B x * `` if batch_first = False
+        Variable of size ``B x T x * `` otherwise
+    """
+
+    if len(lengths) != len(sequences):
+        raise ValueError("number of elements in lengths and sequences didn't match")
+
+    # assuming trailing dimensions and type of all the Variables
+    # in sequences are same and fetching those from sequences[0]
+    max_len = lengths[0]
+    trailing_dims = list(sequences[0].size())[1:]
+    prev_l = max_len
+    if batch_first:
+        out_dims = [len(sequences), max_len] + trailing_dims
+        out_variable = Variable(sequences[0].data.new(*out_dims).zero_())
+        for i, variable, length in zip(range(len(lengths)), sequences, lengths):
+            # temperory sort check, can be removed when we handle sorting internally
+            if prev_l < length:
+                    raise ValueError("lengths array has to be sorted in decreasing order")
+            prev_l = length
+            if length < max_len:
+                prev_l = length
+                padding_dims = [lengths[0] - length] + trailing_dims
+                filler = Variable(variable.data.new(*padding_dims).zero_())
+                out_variable[i] = torch.cat((variable, filler))
+            else:
+                out_variable[i] = variable
+    else:
+        # repeating the same logic but with T as first dimension
```

Comment:
I believe you can remove this code duplication. For the indexing, you could use `.select` and specify the dimension to index in the condition. What do you think?

Reply:
I was looking for something similar but couldn't find anything. I guess ``.select`` makes a copy and I won't be able to assign the value to that slice. Am I not correct?
for ex. below case my ``a`` won't change at all
``
>>> a = Variable(torch.rand(2,3))
>>> a.select(1, 1).data = torch.zeros(2)
``

Reply:
`select` does not perform a copy, but returns an object that shares the same storage. But maybe using `select` might become cumbersome.
If it is easier, you could instead index with tuples (which was used internally in an old version of `select`).
Something like
```python
a = Variable(torch.rand(2, 3))
index = (i, ) if batch_first else (slice(None), i)
a[index] = ...
```

Reply:
Yup, there should be a single code path for both cases. There's no reason to separate them. I think it's best to use `select`.